### ISSUE TITLE
nodePool for ARM in matrix not recognized

### ISSUE THREAD

**Issue by junaruga** [Shippable/support#4608]
### Description of your issue:

`nodePool`: `shippable_shared_aarch32` and `shippable_shared_aarch64` in `matrix` are not recognized.
But it seems that `default_node_pool` is recognized, seeing theerror message.

```
$ cat shippable.yml
language: ruby
matrix:
  include:
    - rvm: 2.5
      nodePool: default_node_pool
    - rvm: 2.5
      nodePool: shippable_shared_aarch32
    - rvm: 2.5
      nodePool: shippable_shared_aarch64
```

The error message on the console page.

```
Invalid nodePool: shippable_shared_aarch32 specified in matrix include section.Invalid nodePool: shippable_shared_aarch64 specified in matrix include section.
```

I have the available nodes.

![shippable_node_pools2](https://user-images.githubusercontent.com/121989/47096499-9c289000-d22f-11e8-89ff-7b514901f548.png)

- Link to the build failure
  https://app.shippable.com/github/junaruga/ruby-pg/runs/37/1/console
- Scripts called from the yml (optional)
  https://github.com/junaruga/ruby-pg/blob/feature/ci-32bit-shippable-report-matrix-nodePool/shippable.yml

**Comment by manishas** [comment]
@ric03uec can we take a look at this?

**Comment by ric03uec** [comment]
@junaruga:
- for the yml you've posted, you don't really need to use `matrix ` block. the following yml snippet should work
```
language: ruby
runtime:
    - nodePool: default_node_pool
    - nodePool: shippable_shared_aarch32
    - nodePool: shippable_shared_aarch64
```
This will generate three builds that run on the specified node pools.
HTH

**Comment by junaruga** [comment]
@ric03uec  The reason why I did not use `runtime/nodePool` is because I wanted to use the `env` together like this. `runtime/nodePool` supporting `env`?

```
language: ruby
matrix:
  include:
    - rvm: 2.4 
      nodePool: default_node_pool
      env: PGVERSION=10.0-1-linux-64
    - rvm: 2.4 
      nodePool: shippable_shared_aarch32
      env: PGVERSION=10.0-1-linux
    - rvm: 2.5 
      nodePool: shippable_shared_aarch64
      env: PGVERSION=10.0-1-linux-64
```

See https://github.com/junaruga/ruby-pg/blob/feature/ci-32bit-shippable-report-matrix-nodePool/shippable.yml

**Comment by junaruga** [comment]
Maybe current logic to detect the target node pool list to start is to check only `runtime/nodePool` list element.
Shippable needs to check both `runtime/nodePool` and `matrix/include/nodePool` list elements.

Because when I run below experimental config file, 6 cases are run.
https://app.shippable.com/github/junaruga/ruby-pg/runs/41/summary/console
https://github.com/junaruga/ruby-pg/blob/feature/ci-32bit-shippable-report-matrix-nodePool-2/shippable.yml

1. runtime/nodePool: default_node_pool
2. runtime/nodePool: shippable_shared_aarch32
3. runtime/nodePool: shippable_shared_aarch64
4. matrix/include/runtime/nodePool: default_node_pool
5. matrix/include/runtime/nodePool: shippable_shared_aarch32
6. matrix/include/runtime/nodePool: shippable_shared_aarch64

And case 5 and 6 are started without error.

For case 5. `shippable_shared_aarch32` is recognized.

```
Provision job node:
Running job on shared node pool: shippable_shared_aarch32
```

But the result of `uname -a` is `aarch64`. It looks wrong. It should be `aarch32`. This is an another issue.

```
uname -a
Linux 125ae9fa41ba 4.15.0-34-generic #37~16.04.1-Ubuntu SMP Tue Aug 28 10:44:16 UTC 2018 aarch64 aarch64 aarch64 GNU/Linux
```

**Comment by ric03uec** [comment]
@junaruga
- the matrix->include part of the yml does not seem to be working as expected. You're right in pointing out that the `matric/include/nodePool` should be checked after figuring out the node pools. I was able to reproduce this with a similar yml setup that ran 6 builds (it should've run only 3). We'll fix this issue at our end by next release.
In the meanwhile, you can use a workaround to get your jobs unblocked. This will involve configuring a `build` section for each node pool. The downside will be duplication of your CI code but once we fix the issue, you can remove the duplicated code with the yml setup you currently have. Can you configure your job with a yml like following?

```
runtime:
  - nodePool: shippable_shared_aarch32
    build:
      ci:
        - export PGVERSION=10.0-1-linux
        - echo "Add other CI commands"
  - nodePool: dynamic_u16_xlarge
    build:
      ci:
        - export PGVERSION=10.0-1-linux-64
        - echo "Add other CI commands"
  - nodePool: dynamic_u14_xlarge
    build:
      ci:
        - export PGVERSION=10.0-1-linux-64
        - echo "Add other CI commands"
```

- for aarch32 nodes, the machines that we use are aarch64 hosts(on packet.net) that are capable of running programs meant for aarch32 architecture. Because of this, the `uname -a` show `aarch64` instead of `aarch32`.
- i'll keep you posted on the progress of the matrix includes issue. let me know if the suggested workaround moves things ahead.

**Comment by junaruga** [comment]
Thanks for the checking.
I tried your way.
But it seems that `runtime/build` does not support `rvm`, right?

```
runtime:
  - nodePool: shippable_shared_aarch32
    rvm: 2.5
    build:
      ci:
```

**Comment by ric03uec** [comment]
yes, `rvm` is not supported inside `runtime/build` because `rvm` is set at the topmost level while constructing the job matrix. Here's another yml that will work for you 

```
language: ruby

rvm:
 - 2.5

env:
- PGVERSION=10.0-1-linux-64

runtime:
- nodePool: dynamic_u14_xlarge
- nodePool: shippable_shared_aarch32
- nodePool: dynamic_u16_xlarge

matrix:
  exclude:
    - rvm: 2.5
      nodePool: shippable_shared_aarch32
      env: PGVERSION=10.0-1-linux-64
  include:
    - rvm: 2.5
      nodePool: shippable_shared_aarch32
      env: PGVERSION=10.0-1-linux    
   
build:
 ci:
   - |
    echo "PGVERSION: ${PGVERSION}"
```
- This will create a 3 build matrix and you wont have to repeat the `build->ci` block for the different `runtime` versions. The logic being executed here is
    - construct a job matrix from `rvm` list (1) x `runtime version` (3) = 3
    - exclude the value `rvm: 2.5` and `nodePool: shippable_shared_aarch32`. Total jobs = 2
    - include the value `rvm: 2.5` and `nodePool: shippable_shared_aarch32` with new env. Total jobs = 3
- I've tested this to be working correctly. if you have a different scenario, i can help you build the relevant yml

**Comment by junaruga** [comment]
Thank you for sharing the way.

**Comment by ric03uec** [comment]
@junaruga we've fixed the issues you were facing with composing the yml for multiple node pools. the following structure will work for you now

```
language: ruby
  
#services:
#  - postgres   
matrix:
  include:
    - rvm: 2.4 
      nodePool: dynamic_u16_xlarge
      env: PGVERSION=10.0-1-linux-64
    - rvm: 2.4 
      nodePool: shippable_shared_aarch32
      env: PGVERSION=10.0-1-linux
    - rvm: 2.5 
      nodePool: shippable_shared_aarch64
      env: PGVERSION=10.0-1-linux-64
```

- this will generate 3 builds on each node pool with the specified env and correct ruby version.

**Comment by ric03uec** [comment]
since this is working now, i'm closing the issue. 
@junaruga please feel free to re-open if you run into any problems.

**Comment by junaruga** [comment]
@ric03uec thank you! I appreciate it!
Sorry. As I was busy for another task, I did not have a time to verity it. But I will use and check it later!

**Comment by junaruga** [comment]
@ric03uec thank you! I checked it now. It works!
